### ISSUE TITLE
SDK fails on Google App Engine

### ISSUE THREAD

**Issue by stickfigure** [paypal/PayPal-Java-SDK#104]
When I try to run this code on Google App Engine (in the dev environment)....

```java
	@GET
	@Path("/wtf")
	public String wtf() throws PayPalRESTException {
		Properties props = new Properties();
		props.setProperty("http.GoogleAppEngine", "true");
		props.setProperty("service.EndPoint", "https://api.sandbox.paypal.com");
		props.setProperty("clientId", "...");
		props.setProperty("clientSecret", "...");

		final OAuthTokenCredential oAuthTokenCredential = Payment.initConfig(props);

		return oAuthTokenCredential.getAccessToken();
	}
```

I get this exception:

```
Caused by: java.lang.ExceptionInInitializerError
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:190)
	at com.google.appengine.tools.development.agent.runtime.RuntimeHelper.checkRestricted(RuntimeHelper.java:70)
	at com.google.appengine.tools.development.agent.runtime.Runtime.checkRestricted(Runtime.java:65)
	at com.paypal.base.DefaultHttpConnection.<init>(DefaultHttpConnection.java:35)
	at com.paypal.base.ConnectionManager.getConnection(ConnectionManager.java:37)
	at com.paypal.base.rest.OAuthTokenCredential.generateOAuthToken(OAuthTokenCredential.java:194)
	at com.paypal.base.rest.OAuthTokenCredential.generateAccessToken(OAuthTokenCredential.java:165)
	at com.paypal.base.rest.OAuthTokenCredential.getAccessToken(OAuthTokenCredential.java:133)
	at com.example.Development.wtf(Development.java:134)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.google.appengine.tools.development.agent.runtime.Runtime.invoke(Runtime.java:130)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:137)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:296)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:250)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:237)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:356)
	... 43 more
Caused by: java.security.AccessControlException: access denied ("java.io.FilePermission" "/Users/jeff/.m2/repository/com/google/appengine/appengine-java-sdk/1.9.21/appengine-java-sdk/appengine-java-sdk-1.9.21/lib/appengine-tools-api.jar/sdk_config.properties" "read")
	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:372)
	at java.security.AccessController.checkPermission(AccessController.java:559)
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
	at com.google.appengine.tools.development.DevAppServerFactory$CustomSecurityManager.checkPermission(DevAppServerFactory.java:429)
	at java.lang.SecurityManager.checkRead(SecurityManager.java:888)
	at java.io.File.exists(File.java:808)
	at com.paypal.base.util.ResourceLoader.searchDirectories(ResourceLoader.java:75)
	at com.paypal.base.util.ResourceLoader.searchClasspath(ResourceLoader.java:67)
	at com.paypal.base.util.ResourceLoader.getInputStream(ResourceLoader.java:50)
	at com.paypal.base.ConfigManager.<init>(ConfigManager.java:92)
	at com.paypal.base.ConfigManager.getInstance(ConfigManager.java:110)
	at com.paypal.base.SSLUtil.<clinit>(SSLUtil.java:47)
	... 63 more```

**Comment by juwlee** [comment]
Hi @stickfigure , it seems you are trying to set config values dynamically from your java application. Looking at the stack trace, I think you are using an old version. Can you please try the latest version? There have been a few changes that may fix the issue.

**Comment by stickfigure** [comment]
Thanks for the response.

I upgraded from 1.2.3 to 1.2.4 (the latest rest-api-sdk in maven central) and still get the same error:

```
java.security.AccessControlException: access denied ("java.io.FilePermission" "/Users/jeff/.m2/repository/com/google/appengine/appengine-java-sdk/1.9.21/appengine-java-sdk/appengine-java-sdk-1.9.21/lib/appengine-tools-api.jar/sdk_config.properties" "read")
	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:372)
	at java.security.AccessController.checkPermission(AccessController.java:559)
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
	at com.google.appengine.tools.development.DevAppServerFactory$CustomSecurityManager.checkPermission(DevAppServerFactory.java:429)
	at java.lang.SecurityManager.checkRead(SecurityManager.java:888)
	at java.io.File.exists(File.java:808)
	at com.paypal.base.util.ResourceLoader.searchDirectories(ResourceLoader.java:75)
	at com.paypal.base.util.ResourceLoader.searchClasspath(ResourceLoader.java:67)
	at com.paypal.base.util.ResourceLoader.getInputStream(ResourceLoader.java:50)
	at com.paypal.base.ConfigManager.<init>(ConfigManager.java:94)
	at com.paypal.base.ConfigManager.getInstance(ConfigManager.java:112)
	at com.paypal.base.SSLUtil.<clinit>(SSLUtil.java:65)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:190)
	at com.google.appengine.tools.development.agent.runtime.RuntimeHelper.checkRestricted(RuntimeHelper.java:70)
	at com.google.appengine.tools.development.agent.runtime.Runtime.checkRestricted(Runtime.java:65)
	at com.paypal.base.DefaultHttpConnection.<init>(DefaultHttpConnection.java:35)
	at com.paypal.base.ConnectionManager.getConnection(ConnectionManager.java:37)
	at com.paypal.base.rest.OAuthTokenCredential.generateOAuthToken(OAuthTokenCredential.java:194)
	at com.paypal.base.rest.OAuthTokenCredential.generateAccessToken(OAuthTokenCredential.java:165)
	at com.paypal.base.rest.OAuthTokenCredential.getAccessToken(OAuthTokenCredential.java:133)
	at com.example.Development.wtf(Development.java:134)
```

Yes, I am loading the configuration dynamically. My dev env uses different credentials than prod, etc.

**Comment by stickfigure** [comment]
Sorry to bug you guys but is there any chance you can suggest a workaround for this issue? I'm confused by why the SDK is trying to load `sdk_config.properties` when I am passing the properties in dynamically. Thanks.

**Comment by dormitionskete** [comment]
Well, if you want a work-around, you could find where it is trying to load the sdk_config.properties, and modify that method to dynamically load your properties.

**Comment by juwlee** [comment]
I have a fix that I am planning on releasing soon, along with some other fixes. To share the details, `AccessControlException` is not caught by [the GAE check logic](https://github.com/paypal/PayPal-Java-SDK/blob/master/rest-api-sdk/src/main/java/com/paypal/base/ConfigManager.java#L96) as it is currently looking for `IOException`, which is not part of `AccessControlException` inheritance chain.

@stickfigure It would be great if you could test it out once released. I will keep you updated on the release.

**Comment by stickfigure** [comment]
I'll happily build and test a snapshot as soon as the fix is committed :-)

**Comment by juwlee** [comment]
Thanks @stickfigure. It has been pushed to branch [v1.2.5](https://github.com/paypal/PayPal-Java-SDK/tree/v1.2.5). Please try it out and let me know.

**Comment by stickfigure** [comment]
It passes my smoke test. Thanks!

**Comment by juwlee** [comment]
Great! I will bundle the fix with others in one release and close this issue when it is out.

**Comment by juwlee** [comment]
Fix released in `v1.2.5`. Thanks for reporting @stickfigure .
